Blaze: Add UI for budget setting screen

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #11599 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds a new screen for setting up the budget for a Blaze campaign. This includes only the UI logic and budget calculation for the screen. Fetching and displaying estimated impressions will be added in a separate PR. Unit tests will also be added later.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Log in to a store with at least one existing product and is eligible for Blaze.
- Navigate to the Products tab and select any public product.
- On the product form, select Promote with Blaze.
- Notice that the campaign creation form is presented.
- Select the budget field, notice that a new screen for budget setting is displayed.
- Tap the estimated impression text, a new sheet should be displayed for more information on impressions.
- Tap the Edit button on the date range to change the duration and start date. The formatted duration should be updated correctly.
- Update the budget with the budget slider, the formatted amount should be updated correctly.
- Tap Update button, the details should be updated correctly on the budget field of the campaign creation form.

📝 There's still an ongoing discussion regarding formatting the budget amount on Figma - I'll update the format later after we reach a conclusion.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-ios/assets/5533851/48a4a3ae-4bcf-40e5-8850-6d4bf4db9de8



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Add file for BlazeBudgetSettingView

itsmeichigo committed:
Update main layout for budget setting screen

itsmeichigo committed:
Add localized text for budget setting screen

itsmeichigo committed:
Add view model for budget setting view

itsmeichigo committed:
Navigate to budget setting from campaign creation form

itsmeichigo committed:
Add impression info view

itsmeichigo committed:
Update padding for impression info view

itsmeichigo committed:
Silence hound

itsmeichigo committed:
Add duration setting view

itsmeichigo committed:
Add localized string for duration setting screen

itsmeichigo committed:
Format amount

itsmeichigo committed:
Format day count for duration setting

itsmeichigo committed:
Update day count slider range and step

itsmeichigo committed:
Add total spend label

itsmeichigo committed:
Format date range on budget setting view

itsmeichigo committed:
Handle completion for budget settings

itsmeichigo committed:
Format detail for budget field on campaign creation screen

itsmeichigo committed:
Add more comments for the SwiftUI view

hafizrahman commented on WooCommerce/Classes/ViewRelated/Blaze/BudgetSetting/BlazeBudgetSettingView.swift:
```diff
@@ -0,0 +1,301 @@
+        }
+        .sheet(isPresented: $showingDurationSetting) {
+            if #available(iOS 16, *) {
+                durationSettingView.presentationDetents([.medium, .large])
```
Design discussion:

When the sheet is pulled to show larger detent here, it leaves quite a big empty space that does nothing. This is with an already large font:


https://github.com/woocommerce/woocommerce-ios/assets/266376/1cdbe415-add1-45f5-9ad6-6724956b611c

With smaller / more normal fonts, it looks more sparse. I wonder if it's needed, or it's fine to leave it this way?

itsmeichigo replied:
Good point! I added a change in abd547607db8e435c190249e6f27331d14de9fed to decide the presentation detent depending on size category. If an accessibility font is being used, large detent is added to prepare for the worst case.

hafizrahman commented on WooCommerce/Classes/ViewRelated/Blaze/BudgetSetting/BlazeBudgetSettingView.swift:
```diff
@@ -0,0 +1,301 @@
+        .padding(.top)
+    }
+
+    var durationSettingView: some View {
```
In landscape mode I can't seem to be able to dismiss this view:

https://github.com/woocommerce/woocommerce-ios/assets/266376/4ce46518-f830-4f22-9ede-4053252ef0ea

itsmeichigo replied:
This can still be dismissed by tapping Apply. But good point about no Dismiss button - I consulted Ann on Figma and will update after getting a confirmation from her.

hafizrahman replied:
In portrait mode I can drag down the sheet to make it disappear. That's the same action I wanted to do here in landscape mode, but the problem seems like the dragging bar at the top is not shown in landscape 🤔

itsmeichigo replied:
This is the behavior for iOS - the partial modal is only available if the vertical size class is not compact. Ann confirmed on Figma to add a button to dismiss this view, so I updated that in 5478da960984de3a6f5a3a31ef111aba74e2fe19. To simplify the alignment of the title, I use a navigation bar so there's a divider below the bar as the default behavior.

hafizrahman commented on WooCommerce/Classes/ViewRelated/Blaze/BudgetSetting/BlazeBudgetSettingView.swift:
```diff
@@ -0,0 +1,301 @@
+
+                    Spacer()
+
+                    DatePicker(selection: $viewModel.startDate, displayedComponents: [.date]) {
```
Can this be made to prevent selecting past dates? I think it makes no sense to make a campaign starting in the past.

itsmeichigo replied:
Nice catch! Added the date range and sneaked in a layout fix in b09338a4712d83c6d31b058fc0784b9530cae0cc.

hafizrahman reviewed (commented):
This generally works well, nice work.

I added some comments and questions.

itsmeichigo committed:
Allow picking only date from present for campaign start date

itsmeichigo committed:
Update presentation detent depending on size category

itsmeichigo committed:
Show USD in total budget amount

itsmeichigo committed:
Keep sizeCategory private

itsmeichigo commented:
Thanks, Hafiz for the reviews! I responded to your comments and addressed the issues in the subsequent commits. I also added a change to include USD in the total budget following the latest design discussion. This PR is ready for another look.

itsmeichigo committed:
Add cancel button to dismiss duration setting view

itsmeichigo committed:
Remove top padding for impression view

itsmeichigo committed:
Merge branch 'trunk' into feat/11599-budget-setting-ui

itsmeichigo committed:
Fix build failure

hafizrahman reviewed (approved):
Thanks for the discussions and working on the changes. I think this is in a good state now :shipit:
